File: go/mechanisms/evm/batched/client/permit2.go
package client

import (
	"context"
	"fmt"
	"math/big"
	"time"

	"github.com/x402-foundation/x402/go/mechanisms/evm"
	"github.com/x402-foundation/x402/go/mechanisms/evm/batched"
	"github.com/x402-foundation/x402/go/types"
)

// CreateBatchedPermit2DepositPayload builds a deposit + voucher payload that
// funds the channel via the universal Permit2 contract using a channel-bound
// `PermitWitnessTransferFrom` authorization. The witness binds the transfer to
// the derived channelId so the Permit2DepositCollector can verify which
// channel the funds belong to.
//
// Mirrors TS `createBatchSettlementPermit2DepositPayload`.
func CreateBatchedPermit2DepositPayload(
	ctx context.Context,
	signer evm.ClientEvmSigner,
	requirements types.PaymentRequirements,
	channelConfig batched.ChannelConfig,
	depositAmount string,
	maxClaimableAmount string,
	voucherSigner evm.ClientEvmSigner,
) (types.PaymentPayload, error) {
	networkStr := string(requirements.Network)

	chainId, err := evm.GetEvmChainId(networkStr)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to get chain ID: %w", err)
	}

	depositBig, ok := new(big.Int).SetString(depositAmount, 10)
	if !ok {
		return types.PaymentPayload{}, fmt.Errorf("invalid deposit amount: %s", depositAmount)
	}

	nonce, err := evm.CreatePermit2Nonce()
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to create permit2 nonce: %w", err)
	}
	nonceBig, ok := new(big.Int).SetString(nonce, 10)
	if !ok {
		return types.PaymentPayload{}, fmt.Errorf("invalid permit2 nonce: %s", nonce)
	}

	deadline := fmt.Sprintf("%d", time.Now().Unix()+int64(requirements.MaxTimeoutSeconds))
	deadlineBig, ok := new(big.Int).SetString(deadline, 10)
	if !ok {
		return types.PaymentPayload{}, fmt.Errorf("invalid permit2 deadline: %s", deadline)
	}

	channelId, err := batched.ComputeChannelId(channelConfig, chainId)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to compute channel ID: %w", err)
	}
	channelIdBytes, err := evm.HexToBytes(channelId)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to parse channel id: %w", err)
	}

	tokenAddress := evm.NormalizeAddress(requirements.Asset)
	spenderAddress := evm.NormalizeAddress(batched.Permit2DepositCollectorAddress)

	domain := evm.TypedDataDomain{
		Name:              batched.Permit2DomainName,
		ChainID:           chainId,
		VerifyingContract: batched.Permit2Address,
	}

	// EIP-712 type definitions: include EIP712Domain alongside the witness types
	// so the signer can hash the domain separator the same way Permit2 does.
	typedDataTypes := map[string][]evm.TypedDataField{
		"EIP712Domain": {
			{Name: "name", Type: "string"},
			{Name: "chainId", Type: "uint256"},
			{Name: "verifyingContract", Type: "address"},
		},
		"PermitWitnessTransferFrom": batched.BatchPermit2WitnessTypes["PermitWitnessTransferFrom"],
		"TokenPermissions":          batched.BatchPermit2WitnessTypes["TokenPermissions"],
		"DepositWitness":            batched.BatchPermit2WitnessTypes["DepositWitness"],
	}

	message := map[string]interface{}{
		"permitted": map[string]interface{}{
			"token":  tokenAddress,
			"amount": depositBig,
		},
		"spender":  spenderAddress,
		"nonce":    nonceBig,
		"deadline": deadlineBig,
		"witness": map[string]interface{}{
			"channelId": channelIdBytes,
		},
	}

	signature, err := signer.SignTypedData(ctx, domain, typedDataTypes, "PermitWitnessTransferFrom", message)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to sign permit2 authorization: %w", err)
	}

	actualVoucherSigner := signer
	if voucherSigner != nil {
		actualVoucherSigner = voucherSigner
	}
	voucher, err := SignVoucher(ctx, actualVoucherSigner, channelId, maxClaimableAmount, networkStr)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to sign voucher: %w", err)
	}

	depositPayload := &batched.BatchedDepositPayload{
		Type:          "deposit",
		ChannelConfig: channelConfig,
		Voucher:       *voucher,
		Deposit: batched.BatchedDepositData{
			Amount: depositAmount,
			Authorization: batched.BatchedDepositAuthorization{
				Permit2Authorization: &batched.BatchedPermit2Authorization{
					From: signer.Address(),
					Permitted: batched.BatchedPermit2TokenPermissions{
						Token:  tokenAddress,
						Amount: depositAmount,
					},
					Spender:  spenderAddress,
					Nonce:    nonce,
					Deadline: deadline,
					Witness: batched.BatchedPermit2Witness{
						ChannelId: channelId,
					},
					Signature: evm.BytesToHex(signature),
				},
			},
		},
	}

	return types.PaymentPayload{
		X402Version: 2,
		Payload:     depositPayload.ToMap(),
	}, nil
}

File: python/x402/mechanisms/evm/batch_settlement/client/permit2.py
"""Permit2 deposit payload construction (client side)."""

from __future__ import annotations

import time

try:
    from eth_utils import to_checksum_address
except ImportError as e:
    raise ImportError(
        "EVM mechanism requires ethereum packages. Install with: pip install x402[evm]"
    ) from e

from .....schemas import PaymentRequirements
from ....evm.constants import PERMIT2_ADDRESS
from ....evm.signer import ClientEvmSigner
from ...utils import create_permit2_nonce, get_evm_chain_id
from ..constants import BATCH_PERMIT2_WITNESS_TYPES, PERMIT2_DEPOSIT_COLLECTOR_ADDRESS
from ..types import (
    ChannelConfig,
    DepositAuthorization,
    DepositFields,
    DepositPayload,
    Permit2Authorization,
    Permit2DepositWitness,
    Permit2TokenPermissions,
)
from ..utils import coerce_bytes32, compute_channel_id
from .voucher import sign_voucher


def create_batch_settlement_permit2_deposit_payload(
    signer: ClientEvmSigner,
    requirements: PaymentRequirements,
    channel_config: ChannelConfig,
    deposit_amount: str | int,
    max_claimable_amount: str | int,
    voucher_signer: ClientEvmSigner | None = None,
) -> DepositPayload:
    """Create a deposit payload bundling a Permit2 witness transfer + voucher."""
    chain_id = get_evm_chain_id(str(requirements.network))
    nonce = create_permit2_nonce()
    deadline = str(int(time.time()) + int(requirements.max_timeout_seconds))
    channel_id = compute_channel_id(channel_config, str(requirements.network))

    token = to_checksum_address(requirements.asset)
    spender = to_checksum_address(PERMIT2_DEPOSIT_COLLECTOR_ADDRESS)
    from_addr = to_checksum_address(signer.address)

    domain = {
        "name": "Permit2",
        "chainId": chain_id,
        "verifyingContract": PERMIT2_ADDRESS,
    }
    message = {
        "permitted": {
            "token": token,
            "amount": int(deposit_amount),
        },
        "spender": spender,
        "nonce": int(nonce),
        "deadline": int(deadline),
        "witness": {
            "channelId": coerce_bytes32(channel_id),
        },
    }
    sig_bytes = signer.sign_typed_data(
        domain, BATCH_PERMIT2_WITNESS_TYPES, "PermitWitnessTransferFrom", message
    )
    signature = "0x" + sig_bytes.hex() if not sig_bytes.hex().startswith("0x") else sig_bytes.hex()

    voucher = sign_voucher(
        voucher_signer or signer,
        channel_id,
        max_claimable_amount,
        str(requirements.network),
    )

    payload = DepositPayload()
    payload.channel_config = channel_config
    payload.voucher = voucher
    payload.deposit = DepositFields(
        amount=str(int(deposit_amount)),
        authorization=DepositAuthorization(
            permit2_authorization=Permit2Authorization(
                from_address=from_addr,
                permitted=Permit2TokenPermissions(token=token, amount=str(int(deposit_amount))),
                spender=spender,
                nonce=nonce,
                deadline=deadline,
                witness=Permit2DepositWitness(channel_id=channel_id),
                signature=signature,
            ),
        ),
    )
    return payload


__all__ = ["create_batch_settlement_permit2_deposit_payload"]
